### ISSUE TITLE
Category keys that are integers can be auto-populated

### ISSUE THREAD

**Issue by brindakv** [informatics-isi-edu/protein-database#7]
In mmCIF, some tables have natural keys (such as user-defined protein names), while others have artificial keys (unique row numbers within a table). The artificial keys can be auto-populated when adding a row to a table. 

A related issue is that these artificial keys are redundant since the catalog includes unique row ids for each entry. We could either remove these columns from the table and use the catalog row ids or retain the redundant columns in order to match with the mmCIF data model.

**Comment by svoinea** [comment]
Can you please provide a list with tables and their columns that can be auto-populated?

**Comment by hongsudt** [comment]
hide artificial keys; use RID for chaise presentation. 
Export should use these artificial keys, instead of RID.

**Comment by carlkesselman** [comment]
This seems correct to me.

As a general rule of thumb, if we think the solution requires a trigger…. think again.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Jun 11, 2019, at 6:05 PM, Hongsuda <notifications@github.com<mailto:notifications@github.com>> wrote:


hide artificial keys; use RID for chaise presentation.
Export should use these artificial keys, instead of RID.

**Comment by svoinea** [comment]
The artificial keys are `not null`. I don't see how we can avoid having a trigger for generating a value. By hiding them, they will come with a `NULL` value.

**Comment by carlkesselman** [comment]
Then we should change the column definition.  We really do not want to use triggers in general and here specifically.

I think we should be fine making this column nullok and not a key given that we have the RID column.

Carl

Sent from my iPhone

On Jun 12, 2019, at 4:19 PM, svoinea <notifications@github.com<mailto:notifications@github.com>> wrote:


The artificial keys are not null. I don't see how we can avoid having a trigger for generating a value. By hiding them, they will come with a NULL value.

**Comment by svoinea** [comment]
I think that the artificial keys need to be specified in the provided schema definition, such that the column will not have the `not null` constraint.

Right now, I identified them by being `unique` and `not` part of a foreign key.

**Comment by carlkesselman** [comment]
Correct.  in our ingest routines, we are making them non-null.  I think we need to specifically exclude these as a special case given that we don’t want that behavior.

Carl

----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Jun 13, 2019, at 4:05 PM, svoinea <notifications@github.com<mailto:notifications@github.com>> wrote:


I think that the artificial keys need to be specified in the provided schema definition, such that the column will not have the not null constraint.

Right now, I identified them by being unique and not part of a foreign key.

**Comment by hongsudt** [comment]
@brindakv can we close this issue?

**Comment by hongsudt** [comment]
We need to classify these id columns into group: 
- id of type int --> can these be auto-generated by a single sequencer? 
- id of type text --> can RID be used or should these be user generated ID? 

We can work on this optimization after we have an MVP.
